lib/api/ldp/api.js

```diff
@@ -0,0 +1,137 @@
+   *
+   *   const acl = new AclChecker({host, fetch, store, ...})
+   *
+   *   const ldp = new LdpHttpHandler({store, acl, host, fetch})
+   *   ```
+   *
+   * @param options.store {LdpFileStore|LdpMemoryStore|LdpQuadStore} storage backend
+   *   (either file system based, or in-memory, or SPARQL based quad store, etc)
+   *
+   * @param options.acl {AclChecker} ACL checker component
+   *
+   * @param options.host {SolidHost} Server config object
+   */
+  constructor (options) {
+    this.store = options.store
+    this.acl = options.acl
+    this.host = options.host
+  }
+
+  /**
+   * Provides an Express handler for LDP API requests.
+   * Authentication has already happened earlier in the middleware stack, stored
+   * in session.
+   * Usage:
+   *   ```
+   *   app.use('/', ldp.handleRequest)
+   *   ```
+   *
+   * @see https://github.com/solid/solid-architecture/blob/master/server/request-flow.md
+   *
+   * @throws {Error} 401 If request not authenticated but resource is non-public
+   *
+   * @throws {Error} 404 If resource is not found, OR if found and request
+   *   is not authorized to access it (default behavior, can be overridden by
+   *   owner of resource)
+   *
+   * @throws {Error} 403 If request is not authorized to access resource and
+   *   user has enabled "Request permission" action
+   *
+   * @throws {Error} 400 If invalid parameters (or error parsing request body,
+   *   for cases like PATCH requests)
+   *
+   * @throws {Error} 409 If PATCH operation results in conflict
+   *
+   * @throws {Error} 406 If no appropriate representation found (for content type)
+   *
+   * @throws {Error} 405 If HTTP method not allowed / not implemented
+   */
+  async handleRequest (req, res, next) {
+    try {
+      const operation = await this.operationFrom(req)
+
+      // check that operation is permitted (throws error if not)
+      const permissions = await this.acl.allow(operation)
+
+      // perform the operation and return a result
+      const result = await operation.perform({store: this.store})
+
+      // write both generic and op-specific headers
+      // note: you need `permissions` for the WAC-Allow header
+      operation.writeHeaders({res, result, permissions})
+
+      operation.writeResponse({res, result})
+    } catch (error) {
+      next(error)
+    }
+  }
+
+  /**
+   * LdpOperation factory method
+   *
+   * @param req {IncomingRequest}
+   *
+   * @returns {Promise<LdpOperation>}
+   */
+  async operationFrom (req) {
+    const options = this.parseOperation(req)
+
+    const Operation = LdpOperation.BY_METHOD[options.method]
+    if (!Operation) {
+      throw new HttpError(405, 'Method not supported')
+    }
+
+    return Operation.from(options)
+  }
+
+  /**
+   * @param req {IncomingRequest}
+   *
+   * @returns {object} Operation constructor params object
+   */
+  parseOperation (req) {
+    const { method, headers, session: { authentication } } = req
```

Comment:
The knowledge about where and how the authentication is stored should probably live in another component.

Reply:
Heh, sure thing. I was only including it here cause it was mentioned in the architecture flow doc.